tcg/tcg-op.c

```diff
@@ -3434,62 +3594,58 @@ static void do_atomic_op_i64(TCGv_i64 ret, TCGv_cap_checked_ptr checked_addr,
-}
-
-GEN_ATOMIC_HELPER(fetch_add, add, 0)
-GEN_ATOMIC_HELPER(fetch_and, and, 0)
-GEN_ATOMIC_HELPER(fetch_or, or, 0)
-GEN_ATOMIC_HELPER(fetch_xor, xor, 0)
-GEN_ATOMIC_HELPER(fetch_smin, smin, 0)
-GEN_ATOMIC_HELPER(fetch_umin, umin, 0)
-GEN_ATOMIC_HELPER(fetch_smax, smax, 0)
-GEN_ATOMIC_HELPER(fetch_umax, umax, 0)
-
-GEN_ATOMIC_HELPER(add_fetch, add, 1)
-GEN_ATOMIC_HELPER(and_fetch, and, 1)
-GEN_ATOMIC_HELPER(or_fetch, or, 1)
-GEN_ATOMIC_HELPER(xor_fetch, xor, 1)
-GEN_ATOMIC_HELPER(smin_fetch, smin, 1)
-GEN_ATOMIC_HELPER(umin_fetch, umin, 1)
-GEN_ATOMIC_HELPER(smax_fetch, smax, 1)
-GEN_ATOMIC_HELPER(umax_fetch, umax, 1)
+#define GEN_ATOMIC_HELPER(NAME, OP, NEW, SIGNED)                               \
+    static void *const table_##NAME[16] = {                                    \
+        [MO_8] = gen_helper_atomic_##NAME##b,                                  \
+        [MO_16 | MO_LE] = gen_helper_atomic_##NAME##w_le,                      \
+        [MO_16 | MO_BE] = gen_helper_atomic_##NAME##w_be,                      \
+        [MO_32 | MO_LE] = gen_helper_atomic_##NAME##l_le,                      \
+        [MO_32 | MO_BE] = gen_helper_atomic_##NAME##l_be,                      \
+        WITH_ATOMIC64([MO_64 | MO_LE] = gen_helper_atomic_##NAME##q_le)        \
+            WITH_ATOMIC64([MO_64 | MO_BE] = gen_helper_atomic_##NAME##q_be)};  \
+    void tcg_gen_atomic_##NAME##_i32(TCGv_i32 ret, TCGv_cap_checked_ptr addr,  \
+                                     TCGv_i32 val, TCGArg idx, MemOp memop)    \
+    {                                                                          \
+        if ((tcg_ctx->tb_cflags & CF_PARALLEL) && !ALL_WRITES_ATOMIC) {        \
+            do_atomic_op_i32(ret, addr, val, idx, memop, table_##NAME);        \
+        } else {                                                               \
+            do_nonatomic_op_i32(ret, addr, val, idx, memop, NEW,               \
+                                tcg_gen_##OP##_i32, SIGNED);                   \
+        }                                                                      \
+    }                                                                          \
+    void tcg_gen_atomic_##NAME##_i64(TCGv_i64 ret, TCGv_cap_checked_ptr addr,  \
+                                     TCGv_i64 val, TCGArg idx, MemOp memop)    \
+    {                                                                          \
+        if ((tcg_ctx->tb_cflags & CF_PARALLEL) && !ALL_WRITES_ATOMIC) {        \
+            do_atomic_op_i64(ret, addr, val, idx, memop, table_##NAME);        \
+        } else {                                                               \
+            do_nonatomic_op_i64(ret, addr, val, idx, memop, NEW,               \
+                                tcg_gen_##OP##_i64, SIGNED);                   \
+        }                                                                      \
+    }
```

Comment:
Please keep the original formatting despite what clang-format wants, this makes comparing rather hard.